File: backend/app/routes/projects.py
from typing import List

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session, selectinload

from ..database import get_db
from .. import models, schemas
from ..auth import get_current_user, check_manager_role

router = APIRouter()


@router.get("/", response_model=List[schemas.ProjectResponse])
def get_projects(
    db: Session = Depends(get_db),
    current_user: models.User = Depends(get_current_user),
):
    """List projects. Managers see all, members see only their assigned projects."""
    if current_user.role.role_name.lower() == "manager":
        return db.query(models.Project).options(selectinload(models.Project.users)).all()
    else:
        return current_user.projects


@router.post("/", response_model=schemas.ProjectResponse, status_code=status.HTTP_201_CREATED)
def create_project(
    project_in: schemas.ProjectCreate,
    db: Session = Depends(get_db),
    current_user: models.User = Depends(check_manager_role),
):
    """Create a new project. Manager only."""
    new_project = models.Project(**project_in.model_dump())
    db.add(new_project)
    db.commit()
    db.refresh(new_project)
    return new_project


@router.get("/{project_id}", response_model=schemas.ProjectResponse)
def get_project(
    project_id: int,
    db: Session = Depends(get_db),
    current_user: models.User = Depends(get_current_user),
):
    """Get a single project by ID."""
    project = db.query(models.Project).filter(models.Project.id == project_id).first()
    if not project:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Project not found")
    return project


@router.put("/{project_id}", response_model=schemas.ProjectResponse)
def update_project(
    project_id: int,
    project_in: schemas.ProjectUpdate,
    db: Session = Depends(get_db),
    current_user: models.User = Depends(check_manager_role),
):
    """Update a project. Manager only."""
    project = db.query(models.Project).filter(models.Project.id == project_id).first()
    if not project:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Project not found")
    
    update_data = project_in.model_dump(exclude_unset=True)
    for key, value in update_data.items():
        setattr(project, key, value)
        
    db.commit()
    db.refresh(project)
    return project


@router.delete("/{project_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_project(
    project_id: int,
    db: Session = Depends(get_db),
    current_user: models.User = Depends(check_manager_role),
):
    """Delete a project. Manager only."""
    project = db.query(models.Project).filter(models.Project.id == project_id).first()
    if not project:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Project not found")
    db.delete(project)
    db.commit()

@router.post("/{project_id}/assign/{user_id}", status_code=status.HTTP_200_OK)
def assign_user(
    project_id: int,
    user_id: int,
    db: Session = Depends(get_db),
    current_user: models.User = Depends(check_manager_role),
):
    project = db.query(models.Project).filter(models.Project.id == project_id).first()
    if not project:
        raise HTTPException(status_code=404, detail="Project not found")
    user = db.query(models.User).filter(models.User.id == user_id).first()
    if not user:
        raise HTTPException(status_code=404, detail="User not found")
    
    if user not in project.users:
        project.users.append(user)
        db.commit()
    
    return {"message": "User assigned to project"}

@router.delete("/{project_id}/assign/{user_id}", status_code=status.HTTP_200_OK)
def unassign_user(
    project_id: int,
    user_id: int,
    db: Session = Depends(get_db),
    current_user: models.User = Depends(check_manager_role),
):
    project = db.query(models.Project).filter(models.Project.id == project_id).first()
    if not project:
        raise HTTPException(status_code=404, detail="Project not found")
    user = db.query(models.User).filter(models.User.id == user_id).first()
    if not user:
        raise HTTPException(status_code=404, detail="User not found")
    
    if user in project.users:
        project.users.remove(user)
        db.commit()
    
    return {"message": "User removed from project"}

File: backend/src/routes/projects.ts
import { Router, Response } from 'express';
import prisma from '../lib/prisma';
import { authenticate, requireProjectManager, AuthRequest } from '../middleware/auth';
import { wsManager } from '../lib/ws';

const router = Router();

/** Returns true if the user is a privileged role (Admin or Project Manager) */
function isPrivilegedUser(req: AuthRequest): boolean {
  const roleName: string = req.user?.role?.role_name ?? '';
  return roleName === 'Administrator' || roleName === 'Project Manager';
}

function parseId(param: string | string[] | undefined): number {
  return parseInt(Array.isArray(param) ? param[0] : (param ?? ''), 10);
}

// GET /api/v1/projects/ — list projects (all for privileged, own for members)
router.get('/', authenticate, async (req: AuthRequest, res: Response): Promise<void> => {
  if (isPrivilegedUser(req)) {
    const projects = await prisma.project.findMany({ include: { users: true } });
    res.json(projects);
  } else {
    const user = await prisma.user.findUnique({
      where: { id: req.user.id },
      include: { projects: { include: { users: true } } }
    });
    res.json(user?.projects || []);
  }
});

// POST /api/v1/projects/ — create project (Project Manager+)
router.post('/', authenticate, requireProjectManager, async (req: AuthRequest, res: Response): Promise<void> => {
  const { project_name, description } = req.body;
  if (!project_name) {
    res.status(400).json({ detail: 'project_name is required' });
    return;
  }
  const project = await prisma.project.create({
    data: { project_name, description }
  });
  res.status(201).json(project);
});

// GET /api/v1/projects/:project_id — get single project
router.get('/:project_id', authenticate, async (req: AuthRequest, res: Response): Promise<void> => {
  const project_id = parseId(req.params.project_id);
  if (isNaN(project_id)) {
    res.status(400).json({ detail: 'Invalid project ID' });
    return;
  }
  const project = await prisma.project.findUnique({ where: { id: project_id } });
  if (!project) {
    res.status(404).json({ detail: 'Project not found' });
    return;
  }
  res.json(project);
});

// PUT /api/v1/projects/:project_id — update project (Project Manager+)
router.put('/:project_id', authenticate, requireProjectManager, async (req: AuthRequest, res: Response): Promise<void> => {
  const project_id = parseId(req.params.project_id);
  const existing = await prisma.project.findUnique({ where: { id: project_id } });
  if (!existing) {
    res.status(404).json({ detail: 'Project not found' });
    return;
  }

  const { project_name, description } = req.body;
  const project = await prisma.project.update({
    where: { id: project_id },
    data: { project_name, description }
  });
  res.json(project);
});

// DELETE /api/v1/projects/:project_id — delete project (Project Manager+)
router.delete('/:project_id', authenticate, requireProjectManager, async (req: AuthRequest, res: Response): Promise<void> => {
  const project_id = parseId(req.params.project_id);
  const existing = await prisma.project.findUnique({ where: { id: project_id } });
  if (!existing) {
    res.status(404).json({ detail: 'Project not found' });
    return;
  }
  await prisma.project.delete({ where: { id: project_id } });
  res.status(204).send();
});

// POST /api/v1/projects/:project_id/assign/:user_id — assign user to project
router.post('/:project_id/assign/:user_id', authenticate, requireProjectManager, async (req: AuthRequest, res: Response): Promise<void> => {
  const project_id = parseId(req.params.project_id);
  const user_id = parseId(req.params.user_id);

  const project = await prisma.project.findUnique({ where: { id: project_id } });
  if (!project) {
    res.status(404).json({ detail: 'Project not found' });
    return;
  }

  const user = await prisma.user.findUnique({ where: { id: user_id } });
  if (!user) {
    res.status(404).json({ detail: 'User not found' });
    return;
  }

  await prisma.project.update({
    where: { id: project_id },
    data: { users: { connect: { id: user_id } } }
  });

  wsManager.sendPersonalMessage({
    type: 'PROJECT_ASSIGNED',
    message: `You were assigned to project: ${project.project_name}`,
    project_id: project.id,
    project_name: project.project_name
  }, user_id);

  res.json({ message: 'User assigned to project' });
});

// DELETE /api/v1/projects/:project_id/assign/:user_id — remove user from project
router.delete('/:project_id/assign/:user_id', authenticate, requireProjectManager, async (req: AuthRequest, res: Response): Promise<void> => {
  const project_id = parseId(req.params.project_id);
  const user_id = parseId(req.params.user_id);

  const project = await prisma.project.findUnique({ where: { id: project_id } });
  if (!project) {
    res.status(404).json({ detail: 'Project not found' });
    return;
  }

  const user = await prisma.user.findUnique({ where: { id: user_id } });
  if (!user) {
    res.status(404).json({ detail: 'User not found' });
    return;
  }

  await prisma.project.update({
    where: { id: project_id },
    data: { users: { disconnect: { id: user_id } } }
  });

  wsManager.sendPersonalMessage({
    type: 'PROJECT_UNASSIGNED',
    message: `You were removed from project: ${project.project_name}`,
    project_id: project.id,
    project_name: project.project_name
  }, user_id);

  res.json({ message: 'User removed from project' });
});

export default router;
